### ISSUE TITLE
[TGL][VP9 Encoder][Multiple Sessions][Crash][Linux] Kernel crash with sample_multi_transcode

### ISSUE THREAD

**Issue by daijh** [Intel-Media-SDK/MediaSDK#2554]
I try to enable multiple threading VP9 encoding on TGL, but it caused kernel crash [i915 kernel driver].
I also reproduced the same issue with *sample_multi_transcode*.

## Environment

**MediaSDK Release**
https://github.com/Intel-Media-SDK/MediaSDK/releases/tag/intel-mediasdk-20.3.0

```bash
lscpu

11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
```

```bash
cat /etc/os-release

NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

```bash
cat /proc/version

Linux version 5.9.15-050915-generic (kernel@sita) (gcc (Ubuntu 10.2.0-13ubuntu1) 10.2.0, GNU ld (GNU Binutils for Ubuntu) 2.35.1) #202012160632 SMP Wed Dec 16 11:53:09 UTC 2020
```
 
```bash
vainfo

libva info: VA-API version 1.9.0
libva info: User environment variable requested driver 'iHD'
libva info: Trying to open /opt/intel/mediasdk/lib64/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_9
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.9 (libva 2.6.0)
vainfo: Driver version: Intel iHD driver for Intel(R) Gen Graphics - 20.3.0 (86ec0b6f)
vainfo: Supported profile and entrypoints
```

**Comment by daijh** [comment]
## Test
The input is 6min vp9 ivf, transcoding by 9 joined vp9 encoding. Below is the test script

### multiple_encoder_test.sh
```source
#!/bin/bash

SAMPLE=/opt/intel/mediasdk/share/mfx/samples/_bin/sample_multi_transcode
LOOP=10

for((i=0;i<${LOOP};i++))
do
    echo "loop..."$i
    ${SAMPLE} -par par-vp9-enc.file
    #${SAMPLE} -par par-avc-enc.file
done

echo "Done"

```

### par-vp9-enc.file
```source
-i::vp9 bbb_720p_vp9.ivf -o::sink -join
-i::source  -o::vp9  /tmp/test0.ivf  -w 1280 -h 720  -join
-i::source  -o::vp9  /tmp/test1.ivf  -w 1280 -h 720  -join
-i::source  -o::vp9  /tmp/test2.ivf  -w 1280 -h 720  -join
-i::source  -o::vp9  /tmp/test3.ivf  -w 1280 -h 720  -join
-i::source  -o::vp9  /tmp/test4.ivf  -w 1280 -h 720  -join
-i::source  -o::vp9  /tmp/test5.ivf  -w 1280 -h 720  -join
-i::source  -o::vp9  /tmp/test6.ivf  -w 1280 -h 720  -join
-i::source  -o::vp9  /tmp/test7.ivf  -w 1280 -h 720  -join
-i::source  -o::vp9  /tmp/test8.ivf  -w 1280 -h 720  -join
```

### log
>loop...0
libva info: VA-API version 1.9.0
libva info: User environment variable requested driver 'iHD'
libva info: Trying to open /opt/intel/mediasdk/lib64/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_9
libva info: va_openDriver() returns 0
Multi Transcoding Sample Version 20.3.0

>Par file is: par-vp9-enc.file

>Session 0:
MFX HARDWARE Session 0 API ver 1.34 parameters:
Input  video: VP9
Output video: To child session

>Session 1:
MFX HARDWARE Session 1 API ver 1.34 parameters:
Input  video: From parent session
Output video: VP9

>Session 2:
MFX HARDWARE Session 2 API ver 1.34 parameters:
Input  video: From parent session
Output video: VP9

>Session 3:
MFX HARDWARE Session 3 API ver 1.34 parameters:
Input  video: From parent session
Output video: VP9

>Session 4:
MFX HARDWARE Session 4 API ver 1.34 parameters:
Input  video: From parent session
Output video: VP9

>Session 5:
MFX HARDWARE Session 5 API ver 1.34 parameters:
Input  video: From parent session
Output video: VP9

>Session 6:
MFX HARDWARE Session 6 API ver 1.34 parameters:
Input  video: From parent session
Output video: VP9

>Session 7:
MFX HARDWARE Session 7 API ver 1.34 parameters:
Input  video: From parent session
Output video: VP9

>Session 8:
MFX HARDWARE Session 8 API ver 1.34 parameters:
Input  video: From parent session
Output video: VP9

>Session 9:
MFX HARDWARE Session 9 API ver 1.34 parameters:
Input  video: From parent session
Output video: VP9

>Pipeline surfaces number (DecPool): 11
Session 0 was joined with other sessions
Session 1 was joined with other sessions
Session 2 was joined with other sessions
Session 3 was joined with other sessions
Session 4 was joined with other sessions
Session 5 was joined with other sessions
Session 6 was joined with other sessions
Session 7 was joined with other sessions
Session 8 was joined with other sessions
Session 9 was joined with other sessions


### dmesg
>[ 1780.102797] watchdog: BUG: soft lockup - CPU#1 stuck for 22s! [sample_multi_tr:3349]
[ 1780.102799] Modules linked in: rfcomm cmac algif_hash algif_skcipher af_alg msr overlay bnep snd_hda_codec_hdmi mousedev joydev snd_hda_codec_realtek snd_hda_codec_generic uvcvideo btusb
 videobuf2_vmalloc btrtl videobuf2_memops btbcm rtsx_usb_ms videobuf2_v4l2 btintel rtsx_usb_sdmmc asix videobuf2_common memstick bluetooth usbnet videodev mii rtsx_usb mc ecdh_generic ecc x
86_pkg_temp_thermal intel_powerclamp coretemp snd_sof_pci snd_sof_intel_byt snd_sof_intel_ipc snd_sof_intel_hda_common snd_soc_hdac_hda snd_sof_xtensa_dsp snd_sof_intel_hda snd_sof snd_hda_
ext_core snd_soc_acpi_intel_match snd_soc_acpi ledtrig_audio snd_soc_core snd_compress ac97_bus snd_pcm_dmaengine snd_hda_intel snd_intel_dspcfg snd_hda_codec snd_hda_core snd_hwdep nls_iso
8859_1 snd_pcm snd_seq_midi snd_seq_midi_event kvm_intel snd_rawmidi kvm crct10dif_pclmul ghash_clmulni_intel hid_sensor_accel_3d aesni_intel intel_rapl_msr i915 snd_seq crypto_simd hid_sen
sor_als hid_sensor_trigger industrialio_triggered_buffer
[ 1780.102809]  iwlmvm cryptd glue_helper kfifo_buf mac80211 hid_sensor_iio_common intel_cstate libarc4 snd_seq_device snd_timer input_leds drm_kms_helper wmi_bmof serio_raw industrialio ef
i_pstore cec iwlwifi snd processor_thermal_device rc_core i2c_algo_bit intel_rapl_common fb_sys_fops syscopyarea sysfillrect soundcore cfg80211 8250_dw hid_multitouch sysimgblt intel_soc_dt
s_iosf mac_hid int3403_thermal int340x_thermal_zone int3400_thermal acpi_thermal_rel intel_hid acpi_pad sparse_keymap acpi_tad sch_fq_codel parport_pc ppdev lp parport drm ip_tables x_table
s autofs4 hid_sensor_custom hid_sensor_hub hid_generic i2c_designware_platform i2c_designware_core crc32_pclmul i2c_i801 nvme i2c_smbus nvme_core intel_lpss_pci thunderbolt intel_lpss idma6
4 xhci_pci virt_dma xhci_pci_renesas wmi i2c_hid hid video backlight pinctrl_tigerlake pinctrl_intel
[ 1780.102820] CPU: 1 PID: 3349 Comm: sample_multi_tr Tainted: G     U            5.9.15-050915-generic #202012160632
[ 1780.102820] Hardware name: Insyde TigerLake/Type2 - Board Product Name1, BIOS TN6V0.9 09/18/2020
[ 1780.102821] RIP: 0010:tasklet_kill+0x4c/0x60
[ 1780.102822] Code: 8d 63 08 f0 48 0f ba 6b 08 00 73 1e e8 1d de b2 00 48 8b 43 08 a8 01 75 f3 f0 49 0f ba 2c 24 00 72 ea 48 8b 43 08 a8 02 74 08 <48> 8b 43 08 a8 02 75 f8 f0 80 63 08 fe 5
b 41 5c 5d c3 66 90 55 48
[ 1780.102823] RSP: 0018:ffffb51700e9fc88 EFLAGS: 00000202
[ 1780.102823] RAX: 0000000000000003 RBX: ffff9965b6d8e3a0 RCX: 0000000000000c9e
[ 1780.102824] RDX: 0000000000000e40 RSI: 0000000000000246 RDI: ffff9965b6d8e3a0
[ 1780.102824] RBP: ffffb51700e9fc98 R08: ffff996618003f00 R09: ffff9965b605da00
[ 1780.102825] R10: 0000000000000035 R11: ffff9965e11a4b6a R12: ffff9965b6d8e3a8
[ 1780.102825] R13: 000000000000001a R14: ffff996615e1dac0 R15: 0000000000000001
[ 1780.102826] FS:  00007f4c97794700(0000) GS:ffff996630840000(0000) knlGS:0000000000000000
[ 1780.102826] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 1780.102827] CR2: 0000559324a900b0 CR3: 000000039692e001 CR4: 0000000000770ee0
[ 1780.102827] PKRU: 55555554
[ 1780.102828] Call Trace:
[ 1780.102848]  intel_engine_flush_submission+0x23/0x80 [i915]
[ 1780.102869]  i915_request_wait+0x130/0x2a0 [i915]
[ 1780.102889]  ? __emit_semaphore_wait+0xe0/0xe0 [i915]
[ 1780.102908]  i915_gem_object_wait_reservation+0xef/0x380 [i915]
[ 1780.102927]  i915_gem_wait_ioctl+0xb0/0x1a0 [i915]
[ 1780.102945]  ? i915_gem_object_wait+0x50/0x50 [i915]
[ 1780.102952]  drm_ioctl_kernel+0xae/0xf0 [drm]
[ 1780.102959]  drm_ioctl+0x245/0x400 [drm]
[ 1780.102977]  ? i915_gem_object_wait+0x50/0x50 [i915]
[ 1780.102978]  ? do_futex+0x160/0x1e0
[ 1780.102979]  ? __x64_sys_futex+0x13d/0x1d0
[ 1780.102980]  __x64_sys_ioctl+0x91/0xc0
[ 1780.102981]  do_syscall_64+0x38/0x90
[ 1780.102982]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[ 1780.102983] RIP: 0033:0x7f4c9e96e50b
[ 1780.102983] Code: 0f 1e fa 48 8b 05 85 39 0d 00 64 c7 00 26 00 00 00 48 c7 c0 ff ff ff ff c3 66 0f 1f 44 00 00 f3 0f 1e fa b8 10 00 00 00 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d 55 3
9 0d 00 f7 d8 64 89 01 48
[ 1780.102984] RSP: 002b:00007f4c977939f8 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[ 1780.102985] RAX: ffffffffffffffda RBX: 00007f4c97793a30 RCX: 00007f4c9e96e50b
[ 1780.102985] RDX: 00007f4c97793a30 RSI: 00000000c010646c RDI: 0000000000000003
[ 1780.102986] RBP: 00000000c010646c R08: 0000000000000000 R09: 0000000000000390
[ 1780.102986] R10: 00007fff3574b080 R11: 0000000000000246 R12: 000055ce4243c360
[ 1780.102986] R13: 0000000000000003 R14: 00007f4c97793af0 R15: 000055ce423f6200

**Comment by daijh** [comment]
Similar kernel crash for h264 multiple threading encoding

### par-avc-enc.file
```source
-i::vp9 bbb_720p_vp9.ivf -o::sink -join
-i::source  -o::h264  /tmp/test0.ivf  -w 1280 -h 720 -join
-i::source  -o::h264  /tmp/test1.ivf  -w 1280 -h 720 -join
-i::source  -o::h264  /tmp/test2.ivf  -w 1280 -h 720 -join
-i::source  -o::h264  /tmp/test3.ivf  -w 1280 -h 720 -join
-i::source  -o::h264  /tmp/test4.ivf  -w 1280 -h 720 -join
-i::source  -o::h264  /tmp/test5.ivf  -w 1280 -h 720 -join
-i::source  -o::h264  /tmp/test6.ivf  -w 1280 -h 720 -join
-i::source  -o::h264  /tmp/test7.ivf  -w 1280 -h 720 -join
-i::source  -o::h264  /tmp/test8.ivf  -w 1280 -h 720 -join
```

### dmesg
>[  216.068034] i915 0000:00:02.0: [drm] Resetting vcs1 for preemption time out
[  216.069967] i915 0000:00:02.0: [drm] GPU HANG: ecode 12:10:28fffffd, in sample_multi_tr [2365]
[  216.069969] GPU hangs can indicate a bug anywhere in the entire gfx stack, including userspace.
[  216.069969] Please file a _new_ bug report at https://gitlab.freedesktop.org/drm/intel/issues/new.
[  216.069970] Please see https://gitlab.freedesktop.org/drm/intel/-/wikis/How-to-file-i915-bugs for details.
[  216.069971] drm/i915 developers can then reassign to the right component if it's not a kernel issue.
[  216.069971] The GPU crash dump is required to analyze GPU hangs, so please always attach it.
[  216.069972] GPU crash dump saved to /sys/class/drm/card0/error

**Comment by dvrogozh** [comment]
I've posted a https://github.com/intel/media-driver/issues/1118#issuecomment-753024815 to derivative issue submitted to driver by @dmitryermilov. Repeating it here below.

Concerning Ubuntu, you can try to use `apt-get install linux-oem` kernel running Ubuntu 20.04.1. One of the delta with default ubuntu kernel is new devices support. Read more here: https://wiki.ubuntu.com/Kernel/OEMKernel. In particular TGL support was added into this kernel for Ubuntu 20.04.1, see https://bugs.launchpad.net/intel/+bug/1855399. And refer to https://wiki.ubuntu.com/FocalFossa/ReleaseNotes/ChangeSummary/20.04.1 (search for TGL).

**Comment by dvrogozh** [comment]
@daijh : if you don't mind, can you, please try running `apt-get install linux-oem` on top of Ubuntu 20.04.1? can you post your observations here?

**Comment by daijh** [comment]
@dvrogozh Thanks for this.

Upgrading oem kernel does not fix the kernel driver crash.

```bash
sudo apt install linux-oem-20.04
sudo reboot

cat /proc/version
>> Linux version 5.6.0-1042-oem (buildd@lcy01-amd64-020) (gcc version 9.3.0 (Ubuntu 9.3.0-17ubuntu1~20.04)) #46-Ubuntu SMP Thu Jan 7 00:51:04 UTC 2021
```

Run the same test scripts, the kernel driver crash.

> [  130.790936] i915 0000:00:02.0: Resetting vcs1 for preemption time out
[  130.792877] i915 0000:00:02.0: GPU HANG: ecode 12:10:28ffff7d, in sample_multi_tr [2426]
[  130.793961] i915 0000:00:02.0: Resetting vcs1 for preemption time out
[  130.795612] i915 0000:00:02.0: GPU HANG: ecode 12:10:28ffff7d, in sample_multi_tr [2426]
[  139.046802] i915 0000:00:02.0: Resetting vcs1 for preemption time out
[  139.048729] i915 0000:00:02.0: GPU HANG: ecode 12:10:28ffff7d, in sample_multi_tr [2426]
[  141.862794] i915 0000:00:02.0: Resetting vcs1 for preemption time out
[  141.864688] i915 0000:00:02.0: GPU HANG: ecode 12:10:28ffff7d, in sample_multi_tr [2426]
[  144.936524] i915 0000:00:02.0: GPU HANG: ecode 12:0:00000000
[  144.936571] i915 0000:00:02.0: Resetting vcs0 for stopped heartbeat on vcs0
[  144.938030] i915 0000:00:02.0: Resetting chip for stopped heartbeat on vcs0
[  144.939403] i915 0000:00:02.0: GPU HANG: ecode 12:0:00000000
[  145.040286] [drm] GuC communication disabled
[  145.040842] [drm:__intel_uc_reset_hw [i915]] *ERROR* Failed to reset GuC, ret = -110
[  145.131327] i915 0000:00:02.0: Failed to reset chip
[  145.235412] i915 0000:00:02.0: sample_multi_tr[2426] context reset due to GPU hang

Can your validation team reproduce it on your golden configuration?
Appreciate if you can share the recommended configuration.

**Comment by dvrogozh** [comment]
@daijh : thank you for trying. That's interesting. And I am especially concerned about a failure to reset a chip (at least GUC). By the way, did the graphics actually recover? For example, were you able to use HW decoding/encoding after the issue w/o system reboot?

One more question: what was the input clip? Is that possible to find out public clip to reproduced? Maybe some clip from ffmpeg fate test set?

**Comment by daijh** [comment]
There is no hard requirement on the  vp9 ivf input clip, just make enough duration (6min in my test) for reproducing the gpu issue. 

I managed to find a public Big_Buck_Bunny 720p WebM and remux into ivf for testing.

```bach
wget https://upload.wikimedia.org/wikipedia/commons/transcoded/c/c0/Big_Buck_Bunny_4K.webm/Big_Buck_Bunny_4K.webm.720p.vp9.webm

ffmpeg -i Big_Buck_Bunny_4K.webm.720p.vp9.webm -an -vcodec copy bbb_720p_vp9.ivf
```

**Comment by daijh** [comment]
- /sys/class/drm/card0/error - i915 error state
[sys-class-drm-card0-error.log](https://github.com/Intel-Media-SDK/MediaSDK/files/5806589/sys-class-drm-card0-error.log)

- full dmesg log
[dmesg.log](https://github.com/Intel-Media-SDK/MediaSDK/files/5806590/dmesg.log)

**Comment by dvrogozh** [comment]
@daijh : cool, thank you. 
@tjaalton : fyi - gpu hangs on tgl on linux-oem kernel

Someone needs to try this on drm-tip kernel and submit issue on FDO if reproduced or figure out which patch in drm-tip fixed the issue.

**Comment by dvrogozh** [comment]
So, I've reproduce this issue on Dell Inspiron 7506 2 in 1 which has 2 gpus: TGL+DG1. As of now:
1. I definitely see the issue on TGL w/ linux-oem kernel. And I have different symptoms. Sometimes system goes completely unresponsive, sometimes graphics stuck, but ssh works. But I was not able to collect dmesg and error state on my own - system not that stable after issue is triggerred. But I most definitely saw this in the log at least once:
```
[ 1780.102797] watchdog: BUG: soft lockup - CPU#1 stuck for 22s! [sample_multi_tr:3349]
```
2. I was not able to reproduce issue on TGL following https://dgpu-docs.intel.com/devices/iris-xe-max-graphics/index.html instruction. However, I might not have waited long enough (but I waited much longer than issue was triggered on TGL)

**NOTE**: to repro issue on TGL you need to boot the kernel with `i915.enable_guc=2`

Now, I have the following observation. I don't actually get what VP9 encoder actually does since from the first glance it is doing something strange. So, with the above parfile which sets the encode in absolutely trivial way (`-o::vp9  /tmp/test0.ivf  -w 1280 -h 720`) and BBB clip obtained via ffmpeg (see https://github.com/Intel-Media-SDK/MediaSDK/issues/2554#issuecomment-759181509), I see this in a **middle** of transcoding:
```
-rw-rw-r--  1 linux linux  161638582 Jan 13 20:56 bbb_720p_vp9.ivf
-rw-rw-r--  1 linux linux  755871744 Jan 14 10:11 test0.ivf
```
So, I am not sure what's the default VP9 encoding mode, but it gives output stream TIMES bigger than the input stream. And that was not finished transcoding. I actually never was able to wait for the 1st iteration in the repro script. I has waited on DG1 for output stream to reach 2GB and it was still running. Someone from VP9 encoding owners should comment on that behavior I think.

**Comment by daijh** [comment]
@dvrogozh The issue is also reproduced for H264 encoding, see par file in https://github.com/Intel-Media-SDK/MediaSDK/issues/2554#issuecomment-752326423

**Comment by dvrogozh** [comment]
@daijh : thanks for paying my attention. I missed that. Well, this just confirms suspicious that big vp9 encoded file size is just something else. Likely bad defaults which trigger cqp or mistake in bitrate calculation. I filed #2569 to track this separately.

Now, I have tried VP9 encoding on drm-tip kernel taken with the tag `intel/CI_DRM_9612` (latest for today) and **I was not able to reproduce the issue on TGL**. This drm-tip kernel seem to roughly correspond to 5.11 or 5.12 kernel. Thus, it seems that there is a fix missing in linux-oem and stock ubuntu kernels. @daijh : can you, please, try drm-tip kernel I have specified on your side to double check my observation?

**Comment by dvrogozh** [comment]
I posted request to https://gitlab.freedesktop.org/drm/intel/-/issues/2937 to help identify patches.

**Comment by tjaalton** [comment]
oem-5.6 is about to be obsolete, hwe-5.8 is what is replacing it, and then there's oem-5.10..

**Comment by daijh** [comment]
## Linux version 5.10.0-1008-oem
- huc failed, so vp9 encoder does not work
- AVC has the same gpu hang issue

```bash
sudo apt install linux-oem-20.04-edge

Linux version 5.10.0-1008-oem (buildd@lgw01-amd64-029) (gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, GNU ld (GNU Binutils for Ubuntu) 2.34) #9-Ubuntu SMP Tue Dec 15 14:22:38 UTC 2020
```

## Linux version 5.11.0-051100rc3-generic
The same as 5.10.0-1008-oem
- huc failed, so vp9 encoder does not work
- AVC has the same gpu hang issue

https://kernel.ubuntu.com/~kernel-ppa/mainline/v5.11-rc3/

```bash
Linux version 5.11.0-051100rc3-generic (kernel@sita) (gcc (Ubuntu 10.2.0-13ubuntu1) 10.2.0, GNU ld (GNU Binutils for Ubuntu) 2.35.1) #202101102331 SMP Sun Jan 10 23:34:29 UTC 2021
```

**Comment by daijh** [comment]
drm-tip head is the next try.

**Comment by dvrogozh** [comment]
I tried vp9 on few kernels from https://kernel.ubuntu.com/~kernel-ppa/mainline/. Results were:
* 5.11.0-051100rc3 - GOOD
* 5.10.7-051007 - GOOD
* 5.9.16-050916 - BAD

I did not try AVC.

When you say "huc failed", what does that mean? Did you install new HUC/GUC FW which new kernels might require.

**Comment by daijh** [comment]
After install new kernel, and reboot. vainfo does not show VP9 encoder.
- Linux version 5.10.0-1008-oem
- Linux version 5.11.0-051100rc3-generic

In my humble opinion, it is general issue existing in all codecs, instead of vp9 specific.
As the same GPU hang issue can be reproduced by H264 on these new kernel, so i don't bother to google how to update firmware.

Can you point me the BKM on firmware update? I might have a try for vp9.
Also, I have trouble to build drm-tip master kernel. Need more try

Linux version 5.11.0-051100rc3-generic is already the latest mainline kernel, can we consider it a drm kernel drive bug?

**Comment by daijh** [comment]
Managed to build drm-tip kernel, the gpu hang still existed for H264 parallel encoding. Did not try VP9 encoding. 

>commit 4cb18d3b417428a4e3834bf93f5de18b57984570 (HEAD -> drm-tip, origin/drm-tip, origin/HEAD)
>Author: Dave Airlie <airlied@redhat.com>
>Date:   Mon Jan 18 11:07:26 2021 +1000
>
>    drm-tip: 2021y-01m-18d-01h-06m-10s UTC integration manifest

**Comment by dvrogozh** [comment]
@daijh : we initially focused on VP9 encoding, not on H264 encoding. Let's avoid mixing experiments since these might very well be different issues! or they might be the same issue... But if we will do different experiments each time, it will be very difficult to root cause the problem. Please, try out drm-tip on your side for VP9 decoder since I clearly see different behavior and can't reproduce issue on drm-tip.

> Can you point me the BKM on firmware update? 

Yes. BKM is the following (see https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git):
```
git clone https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git
cd linux-firmware
sudo cp i915/tgl_* /lib/firmware/i915/
```
So, basically you just copy FW to the required folder. One tricky point (especially if you build kernel on your own) is to copy FW _before_ you install the kernel. Once you loaded a kernel, it's better to check that FW was loaded for your platform. For that look in dmesg:
```
dmesg | grep i915
```

Be aware that VP9 decoder requires HUC FW (and HUC FW requires GUC FW to be loaded). Likely that's why you failed to run VP9 on newer kernel - FW was just missed. Mind also that HUC FW is disabled by default in the kernel, that's why you need to have `i915.enable_guc=2` on kernel cmdline.

**Comment by dvrogozh** [comment]
As for the H264 vs VP9 GPU hangs... We need to evaluate whether that's the same issue or not. If possible, please, provide dmesg logs (better w/ drm.debug=0x1e cmdline option) and GPU hang error states collected on the same kernel for both H264 and VP9 to compare.

**Comment by daijh** [comment]
@dvrogozh Thanks the advice. Sure, let's focus on VP9 encoding. Will try it on my drm-tip.

**Comment by daijh** [comment]
VP9 gpu issue
- drm-tip kernel 
> commit 4cb18d3b417428a4e3834bf93f5de18b57984570 (HEAD -> drm-tip, origin/drm-tip, origin/HEAD)
> Author: Dave Airlie airlied@redhat.com
> Date: Mon Jan 18 11:07:26 2021 +1000
>
> drm-tip: 2021y-01m-18d-01h-06m-10s UTC integration manifest

- firmware
>commit d5288624259300c558480c21a860fcf94187d29d (HEAD -> main, origin/master, origin/main, origin/HEAD)
>Author: Peter Robinson <pbrobinson@gmail.com>
>Date:   Fri Jan 8 11:55:13 2021 +0000



[vp9-dmesg.log](https://github.com/Intel-Media-SDK/MediaSDK/files/5834162/vp9-dmesg.log)
[vp9-sys-class-drm-card0-error.txt](https://github.com/Intel-Media-SDK/MediaSDK/files/5834164/vp9-sys-class-drm-card0-error.txt)

**Comment by dvrogozh** [comment]
@daijh : so, you still see the VP9 hang on drm-tip kernel? Is that right? Which input stream did you use? Was that stream from https://github.com/Intel-Media-SDK/MediaSDK/issues/2554#issuecomment-759181509?

**Comment by daijh** [comment]
Yes, I observed vp9 gpu hang on drm-tip kernel, with exactly the same steps and input clip in previous threads.
But it takes longer time to reproduce. Just let the loop runs 10 times.

**Comment by daijh** [comment]
@dvrogozh Are you able to reproduce the vp9 gpu hang issue on **drm-tip** or ubuntu **Linux version 5.10.0-1008-oem**?

**Comment by dvrogozh** [comment]
@daijh : no. I see pretty stable behavior on drm-tip. I was able to execute your test w/o any issues to the very end. But I saw problem on ubuntu oem kernel as noted before, but I did not try 5.10.0-1008-oem yet (how did you install it by the way?)

I am trying to bisect to the commit on my side which changes behavior. This goes a little slowly.

One question I have to you is: what's the TGL stepping which you are using?

**Comment by daijh** [comment]
@dvrogozh I use below to install Linux version 5.10.0-1008-oem.
```bash
sudo apt install linux-oem-20.04-edge

Linux version 5.10.0-1008-oem (buildd@lgw01-amd64-029) (gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0, GNU ld (GNU Binutils for Ubuntu) 2.34) #9-Ubuntu SMP Tue Dec 15 14:22:38 UTC 2020
```

And lscpu tells **Stepping: 1**
```bash
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   39 bits physical, 48 bits virtual
CPU(s):                          8
On-line CPU(s) list:             0-7
Thread(s) per core:              2
Core(s) per socket:              4
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           140
Model name:                      11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
Stepping:                        1
CPU MHz:                         1200.000
CPU max MHz:                     4200.0000
CPU min MHz:                     400.0000
BogoMIPS:                        4838.40
Virtualization:                  VT-x
L1d cache:                       192 KiB
L1i cache:                       128 KiB
L2 cache:                        5 MiB
L3 cache:                        8 MiB
```

**Comment by daijh** [comment]
Issue is only preproduced on B0 Stepping with special GPU subsilces_mask.
Not reproduced on official GPU.
